### ISSUE TITLE
Multi lingual annotations

### ISSUE THREAD

**Issue by stueynz** [json-schema-org/json-schema-spec#744]
Historically OpenAPI and JSON Schema have worked together to document existing APIs. In a _contract first_ environment they are being asked to do more than merely document existing APIs; they are _defining_ APIs that are yet to be built.  JSON Schema is no longer just about payload validation, the annotations in a JSON Schema complete the definition of the payloads sufficient to define APIs that are yet to be built.

When defining enumeration values, JSON Schema lets us give each enumeration value annotations:

```json
  type: object
  properties:
    Name:
      type: string
    Gender:
       type: string
       oneOf: 
         - const: "M"
           title: "Male"
         - const: "F"
           title: "Female"
         - const: "D"
           title: "Diverse"
           description: "Used where gender known, and is non-binary"
 ```

When JSON Schema is used to _define_ APIs and message payloads, it is not unreasonable for software developers to choose to populate user interface elements (radio buttons, drop down lists, etc.) from the annotation values presented in the JSON schema.  In a multi-lingual environment, we may need multiple annotations, keyed by language.   

Expanding the definition of the `title` (and perhaps `description`) annotations to be either a single string or an object with properties keyed on language tags would be a good idea:

```json
  type: object
  properties:
    Name:
      type: string
    Gender:
       type: string
       oneOf: 
         - const: "M"
           title: {"en-NZ": "Male", "mi-NZ": "Tāne" }
         - const: "F"
           title: {"en-NZ": "Female", "mi-NZ": "Wahine" }
         - const: "D"
           title: {"en-NZ": "Diverse", "mi-NZ": "Ira tāngata kōwhiri kore" }
           description: "Used where gender known, and is non-binary"
 ```

**Comment by awwright** [comment]
@stueynz Is this in response to json-schema-org/json-schema-spec#743 ?

**Comment by stueynz** [comment]
Yes...I was in two minds if it should go on [#743](https://github.com/json-schema-org/json-schema-spec/issues/743) or be a an Issue / Feature request.

**Comment by handrews** [comment]
A less disruptive option might be to follow RFC 8288's approach of `title*`

Keeping individual keywords simple and targeted is a good thing in my view, and there are different mindsets involved in monolingual and multilingual environments.

**Comment by Relequestual** [comment]
My feeling is the same as @johandorland here. Core should be kept simple, but new extended vocabularies can be written.
This will be easier when draft-8 is out, and after such I'm sure myself and others will be happy to help start making a documentation or i11n annotations vocabulary for documentation and UI generation.

**Comment by Relequestual** [comment]
> JSON Schema is no longer just about payload validation

This is exactly why @handrews put so much effort into adding "Vocabularies"! =]

**Comment by stueynz** [comment]
>
> A less disruptive option might be to follow RFC 8288's approach of `title*`
> 

Two things to note about  title* from [rfc-8288](https://tools.ietf.org/html/rfc8288):

- The title* link-param MUST NOT appear more than once in a
   given link-value; - The whole point of what I'm getting at is that we want to be able to put multiple 
   title entries against an enum value, so we'd be departing from rfc-8288 by permitting multiple title* 
   entries;

- rfc-8288 is actually piggybacking off [rfc-8187](https://tools.ietf.org/html/rfc8187) which is trying to 
   encode both language and non US ASCII characters into HTTP headers, which are restricted to only 
   allowing US ASCII characters.

   We don't have the character encoding problem; JSON Schemas are already written in JSON, which 
   includes the implied use of utf-8 encoding; 

We could leave the existing title tag as a single string, and make title* be the object with multiple values keyed by language tags:

```json
 Gender:
       type: string
       oneOf: 
         - const: "M"
           title*: {"en-NZ": "Male", "mi-NZ": "Tāne" }
         - const: "F"
           title*: {"en-NZ": "Female", "mi-NZ": "Wahine" }
         - const: "D"
           title*: {"en-NZ": "Diverse", "mi-NZ": "Ira tāngata kōwhiri kore" }
           description: "Used where gender known, and is non-binary"
```

**Comment by about-code** [comment]
Fully aggree with @johandorland and @Relequestual 

@stueynz To the best of my understanding of REST principles an OpenAPI-Document should first and foremost be treated like a REST resource itself. There is nothing special making it in any way different from the resources it describes. So if some client asks for an OpenAPI document with `Accept-Language: en-NZ` a server may serve an OpenAPI document *fully* localized for `en-NZ` and if a client asks for `en-US` it may get the document for `en-US`. If the language header is missing a server may respond with a default locale (setting the `Content-Language` header). Links to other languages may be represented using Hypermedia.

Just imagine a situation where some of your enum values were permitted for one locale but not applicable to another due to regulatory differences or for whatever reason.  There might be any kind of differences in validation constraints based on a particular locale. After all your OpenAPI descriptions should be in the language of your API customer as well.

I'm afraid trying to nail all those cases with special schema syntax is going to become a half-baked fail because it heavily goes against REST principles IMHO.

**Comment by stueynz** [comment]
I also agree with @johandorland  and @Relequestual  that the correct place do what I'm wanting is in a special Internationalisation vocabulary.  Now that draft-8 is out the door, we can perhaps think about it.

To @about-code I agree that using `Accept-Language` header is how one would query for and receive a fully localised, _mono-lingual_ edition of the JSON schema. 

However:
- Human language is a complex tapestry;  the proposal is to add a vocabulary allowing for the inclusion of _multi-lingual_ annotations in the schema language; because there is a need.

- JSON Schema & OpenAPI are no longer just for the API developers;  our open API spec is the contract to define requirements for systems yet to be built;  we need a mechanism to allow us to specify multiple translations of the field enumerations and descriptions.  Of course we could just slap a few vendor extensions like `x-title`  and `x-description` in as appropriate.
